tests/frontend/test_program.py

```diff
@@ -402,6 +402,52 @@ class DummyCircuit(Compiler):
         with pytest.raises(program.CircuitError, match="program contains 3 modes, but the device 'None' only supports a 2-mode program"):
             new_prog = prog.compile(device=spec, compiler=DummyCircuit())
 
+    @pytest.mark.parametrize(
+        "measure_op", [
+            ops.MeasureFock(),  # MeasureFock
+            ops.MeasureHomodyne(phi=0),  # MeasureX
+            ops.MeasureHomodyne(phi=42),  # MeasureHomodyne
+            ops.MeasureHomodyne(phi=np.pi/2),  # MeasureP
+            ops.MeasureHeterodyne(),  # MeasureHD
+            ops.MeasureHeterodyne(select=0)  # MeasureHeterodyne
+        ],
+    )
+    def test_incorrect_number_of_measurements(self, measure_op):
+        """Test that an exception is raised if the compiler is called with a
+        device spec with an incorrect number of measurements"""
```

Comment:
I added this test here since the other, similar tests were also here (using the same structure), although I think that it might be better to have tests directly for both `assert_number_of_measurements` and `assert_number_of_modes`, instead of implicitly through `prog.compile`.

Perhaps these tests could be updated in a later PR?

Reply:
That seems like a good idea! Would it be worth adding a separate unit test just for `assert_number_of_measurements`?

Reply:
I could do that. In practice, it would do the same thing, just a bit more unit-test-like than this one. I wouldn't mind attempting to refactor some tests at a later stage and moving these tests into an integration tests folder, and making sure that all functions are explicitly tested separately. 

Do you think I should add separate tests for `assert_number_of_measurements` and `assert_number_of_modes` as well? @antalszava

Reply:
From a testing point of view, that would be great because it would constitute a unit test and the test that checks the use of `compile` would be the integration test. Though specifically for these functions, if we deem that testing both "ways" is not too crucial, perhaps we can get away with not having it. Still could be nice to have.

Reply:
Added two extra tests for the assertion functions directly (they're basically the same tests, just avoiding the compilation part). Also added a TODO to the compiler tests.